app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -168,39 +172,119 @@ public function indexAction()
+                'postcode'      => $request_data->zip_code,
+                'country_id'    => $request_data->country_code
+            );
+
+            /** @var Mage_Directory_Model_Country $countryModel */
+            $countryModel = Mage::getModel('directory/country');
+            $countryObj = $countryModel->load($address_data['country_id']);
+            $isRegionAvailable = ($countryObj->getRegionCollection()->getSize() > 0);
+            if (!$isRegionAvailable) {
+                // If country does not have region options for dropdown.
+                $address_data['region'] = $address_data['region_name'];
+            }
+
+            /** @var Mage_Directory_Model_Region $regionModel */
+            $regionModel = Mage::getModel('directory/region');
+            $regionObj= $regionModel->loadByCode($address_data['region'], $address_data['country_id']);
+            $address_data['region_id'] = ($regionObj) ? $regionObj->getId() : null;
+
+            ////////////////////////////////////////////////////////////////////////
+            // Clear previously set estimates.  This helps if this
+            // process fails or is aborted, which will force the actual index action
+            // to get fresh data instead of reading from the session cache
+            ////////////////////////////////////////////////////////////////////////
+            $cacheKeyLocation = $quote->getId() . '_' . $address_data['country_id'];
+            $prefetchCacheKey = $this->generateCacheKey($quote, $address_data);
+            Mage::app()->getCache()->remove('quote_location_' . $cacheKeyLocation);
+            Mage::app()->getCache()->remove($prefetchCacheKey);
+            ////////////////////////////////////////////////////////////////////////
+
+            $quote->getShippingAddress()->addData($address_data);
+            $quote->getBillingAddress()->addData($address_data);
+            try {
+                $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
+            } catch (Exception $e) {
+                $estimate_response = null;
+            }
+
+            $lifeTime = 600; //10 minutes
+            Mage::app()->getCache()->save(
+                serialize($address_data),
+                'quote_location_' . $cacheKeyLocation,
+                array('BOLT_QUOTE_LOCATION'),
+                $lifeTime
+            );
+
+            Mage::app()->getCache()->save(
+                serialize($estimate_response),
+                $prefetchCacheKey,
+                array('BOLT_PREFETCH_SHIPPING_AND_TAX'),
+                $lifeTime
+            );
+        } else {
+            $address_data = array();
+        }
+
+        $response = Mage::helper('core')->jsonEncode(array('address_data' => $address_data));
+        $this->getResponse()->setHeader('Content-type', 'application/json');
+        $this->getResponse()->setHeader('X-Bolt-Cache-Hit', 'HIT');
```

Comment:
this should come to line 147. Idea is to change this based on whether we read it from cache or not (so if cache doesn't hit pleaes call `$this->getResponse()->setHeader('X-Bolt-Cache-Hit', 'MISS');`